pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -76,13 +90,38 @@
     // for connected subscriptions, message expiry will be checked if the backlog is greater than this threshold
     private static final int MINIMUM_BACKLOG_FOR_EXPIRY_CHECK = 1000;
 
+    // Map to keep track of message ack by each txn.
+    private final ConcurrentOpenHashMap<TxnID, ConcurrentOpenHashSet<Position>> pendingAckMessagesMap;
```

Comment:
actually one last comment:

@MarvinCai : can you make `pendingAckMessagesMap` and `pendingAckMessages` non-final and initialize them when they are needed? so we don't need to construct these two structures if people don't use these features, especially when transaction is a feature under development.

Reply:
Sure, changed to only initialize them when client start ack message for transaction, and add  null check for other places using them.